source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -268,11 +319,7 @@ Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream)
 
   if (handshake_complete_) {
     ASSERT(frame_protector_);
-    ENVOY_CONN_LOG(debug, "TSI: protecting buffer size: {}", callbacks_->connection(),
-                   buffer.length());
-    tsi_result status = frame_protector_->protect(buffer, raw_write_buffer_);
-    ENVOY_CONN_LOG(debug, "TSI: protected buffer left: {} result: {}", callbacks_->connection(),
-                   buffer.length(), tsi_result_to_string(status));
+    return repeatProtectAndWrite(buffer, end_stream);
```

Comment:
What about the raw_write_buffer_.length() > 0 code below this early return?

Reply:
I think we still need it to send handshake data to its peer.

Reply:
I agree, but think that it could be moved to the "if (!handshake_complete_) {" branch.

Reply:
Done.